src/compiler/sourcemap.ts

```diff
@@ -51,6 +52,10 @@ namespace ts {
 
         function addSource(fileName: string) {
             enter();
+            const sourceIndexByFileName = fileNameToSourceIndexMap.get(fileName);

```

Comment:
Implementation note: this could alternatively be done without a full cache, using just a single MRU fileName/sourceIndex. That does however require that the second call to `setSourceMapSource` in `emitSourcePos` has to take an additional parameter to not even call `addSource` (which is safe as reverting to the prior source map source is known to have been added) as otherwise 2 MRU items are needed:

https://github.com/microsoft/TypeScript/blob/a7d801176ca6cdfe45162c7d7cc80b2fdd144240/src/compiler/emitter.ts#L5290-L5300

This alternative would also mitigate the issue (I verified that) as it's typical for a single external source map source to be applied on nodes without interference from nodes that have a different external source map source, so a single MRU cache entry is sufficient.